### ISSUE TITLE
[Question] Reproducibility of Results in Mixed Kitchen Task (69.8 ± 0.4).

### ISSUE THREAD

**Issue by InnoMakerQiu** [BayesBrain/Habi#2]
### Description
I am trying to reproduce the results of the Mixed Kitchen task as reported in the paper (69.8 ± 0.4). However, despite following the exact same training process and using the provided configuration files, I am unable to achieve the reported performance. Here are the details of my attempts:

### Attempts and Results
- ​**Training Setup**:
  - Device: 4090s
  - Configuration: Used the original repository's YAML files without any modifications.
  - Training Pipeline:
    1. Trained DV with seed 0.
    2. Trained Habi with seeds 0, 1, and 2.
  - Evaluation: Followed the standard evaluation protocol.

- ​**Results**:
  | Indicator Name           | Mean Reward | Error         |
  |--------------------------|-------------|---------------|
  | habi_d4rl_kitchen_2       | 66.5        | 0.616846821   |
  | habi_d4rl_kitchen_1       | 62.9        | 0.952459973   |
  | habi_d4rl_kitchen_0       | 48.525      | 1.269540516   |
  | habi_d4rl_kitchen_0       | 48.25       | 1.273332243   |

### Request
To help us better understand the discrepancy, could you please provide:
1. The random seeds used for training DV and Habi in the original experiments.
2. If possible, the model weights for DV and Habi used in the original experiments.

Thank you for your support!

**Comment by BayesBrain** [comment]
Hi @InnoMakerQiu . Thank you for your interest in our work.

We re-implemented Habi again on your task using the same code in this repo. The results align quite well with the paper.

> Q1: The random seeds used for training DV and Habi in the original experiments.

We direct use five seeds: 0, 1, 2, 3, 4 without any picking. Our reimplementation performances are as follows:

**Kitchen-Mixed Reimplementation Results**
| Models  | Mean Reward | Error |
|-------|------------|-------|
| seed=0     | 72.7       | 0.3   |
| seed=1     | 67.0       | 0.5   |
| seed=2     | 73.7       | 0.3   |
| seed=3     | 71.8       | 0.4   |
| seed=4     | 69.9       | 0.5   |
| **Average (Reimplement)** | **71.0** |       |
| Habi Paper | 69.8 |       |

**Kitchen-Partial Reimplementation Results**
| Models  | Mean Reward | Error |
|-------|------------|-------|
| seed=0     | 97.5       | 0.5   |
| seed=1     | 97.3       | 0.5   |
| seed=2     | 94.2       | 0.7   |
| seed=3     | 95.1       | 0.7   |
| seed=4     | 91.0       | 0.7   |
| **Average (Reimplement)** | **95.0** |       |
| Habi Paper | 94.8 |       |

It seems that our reimplementations this time are **even a little better than** the paper... 

Are you sure you are following the README, and using the default settings? Please also check the memory usage, and ensure your training is going smoothly. Theoretically, you should be getting similar results.

> Q2: If possible, the model weights for DV and Habi used in the original experiments.

Sure, we have uploaded checkpoints of Habi and also DV (seed=0) on your tasks to Hugging Face via [[this link]](https://huggingface.co/BayesBrain/Habi/tree/main).

You can drag the results folder into `Habi/`, and run with:
```shell
python pipelines/habi_d4rl_kitchen.py task=kitchen-mixed-v0 enable_wandb=0 mode=inference seed=XXX
```

**Comment by InnoMakerQiu** [comment]
Hi @BayesBrain,  

Thank you again for your prompt and detailed support—your provided checkpoints and insights have been immensely helpful!  

### Key Observations & Progress:  
1. **DV Training Discrepancy**:  
   - Using **your DV checkpoint** (seed=0), Habi achieves **74.15 (seed=1)** and **74.3 (seed=0)**, surpassing the paper’s reported 69.8.  
   - However, **our own DV training** only reaches **70.4 ± 0.8**, while your checkpoint yields **74 ± 0.4**, indicating a critical gap in our DV implementation.  

2. **Troubleshooting Steps**:  
   - Verified hyperparameter alignment with YAML configurations.  
   - Testing hardware differences:  
     - Current setup: **4090s (FP32)** vs. original A100 (TF32).  
     - Will retry on **3090s** (supports TF32) and vary DV seeds to isolate hardware/seed impacts.  

### Technical Clarifications:  
1. **Critic Training Duration**  
   - The critic trains for **1M steps** in code/configs, but inference uses the **200k-step checkpoint**.  
   - Is this intentional (e.g., early-stopping best model), or a potential oversight?  

2. **`train_inv` Mode Requirement**  
   - We skipped `train_inv` mode, assuming `separate` inference (via `habi_d4rl_kitchen.py`) uses the policy directly.  
   - Is `train_inv` mandatory for `separate`-mode Habi training, or is omitting it correct?  

Your guidance on these points would greatly help us resolve the discrepancies. We deeply appreciate your time and expertise! 🙌

**Comment by BayesBrain** [comment]
You are welcome. Glad that you get even better performances compared with our paper.

> DV performance

Regarding DV, we recently conducted tests, and its training remains fairly stable (73.725 vs. 73.45). I’m not sure whether the issue lies in your DV training or somewhere in Habi training, but I think the problem is more likely in your DV training process.

> Hardware

Yeah… Personally, I still lean toward the precision differences between V100/A100 and 3090/4090. You might try fixing all precision to 32-bit. However, this repo seems to work quite well on our environment.

> Critic Training Duration. 

Your ideas are good. We also recommend experimenting with different checkpoints for Habi’s policy/critic. It’s possible that in your previous runs, reducing bb_ckpt (early stopping) led to better performance. We didn’t take this approach; instead, we uniformly set it to 1000k, increased the number of seeds, and averaged the results across seeds for consistency, same for all task types.

In theory, with fine-tuned adjustments, you could achieve performance well beyond the reported results in our paper.

> Train_inv

Train_inv might not be useful—you can choose to ignore it.

After all, here's one insight: In the kitchen environment, its offline data is significantly less than in other environments (e.g., antmaze, mujoco). With limited data, task performance fluctuations in individual environments could occur. In this work, we simply increase the seed number and make an average. But it is actually a good research topic in the future.
